common/Geometry2d/Rect.cpp

```diff
@@ -18,6 +18,111 @@ bool Rect::intersects(const Rect& other) const {
     }
 }
 
+int Rect::CohenSutherlandCode(const Point& other) const{
+    const int INSIDE = 0; // 0000
+    const int LEFT = 1;   // 0001
+    const int RIGHT = 2;  // 0010
+    const int BOTTOM = 4; // 0100
+    const int TOP = 8;    // 1000
+    int code;
+    double x,y;
+    x = other.x();
+    y = other.y();
+
+    code = INSIDE;          // initialised as being inside of [[clip window]]
+    if (x < minx())           // to the left of clip window
+        code |= LEFT;
+    else if (x > maxx())      // to the right of clip window
+        code |= RIGHT;
+    if (y < miny())           // below the clip window
+        code |= BOTTOM;
+    else if (y > maxy())      // above the clip window
+        code |= TOP;
+
+    return code;
+}
+
+bool Rect::intersects_(const Segment& other) const{
+    //These constants are needed in here and in the CohenSutherlandCode method - is there
+    //a place where we should make these definitions globally?
+    const int INSIDE = 0; // 0000
+    const int LEFT = 1;   // 0001
+    const int RIGHT = 2;  // 0010
+    const int BOTTOM = 4; // 0100
+    const int TOP = 8;    // 1000
+
+    //Code aggressively borrowed from wikipedia entry Cohen-Sutherland Line Clipping
+    Point p0 = other.pt[0];
+    double x0 = p0.x();
+    double y0 = p0.y();
+    Point p1 = other.pt[1];
+    double x1 = p1.x();
+    double y1 = p1.y();
+
+    int outcode0 = CohenSutherlandCode(p0);
+    int outcode1 = CohenSutherlandCode(p1);
+    bool accept = false;
+
+    while (true) {
```

Comment:
There is no way this can become an infinite loop right? Even with a 0 by 0 rectangle and other possible edge cases?

Reply:
Fixed... GTest cases included in latest commit